File: Intermediate/30.Generate all subsets of a given set using recursion/Solution.py
def generate_subsets(s, index, current_subset):
    # Base case: if we've processed all elements, print the current subset.
    if index == len(s):
        print("{", " ".join(map(str, current_subset)), "}")
        return
    
    # Recurse without including the current element.
    generate_subsets(s, index + 1, current_subset)
    
    # Include the current element.
    current_subset.append(s[index])
    generate_subsets(s, index + 1, current_subset)
    
    # Backtrack: remove the last element before returning.
    current_subset.pop()

if __name__ == "__main__":
    n = int(input("Enter the number of elements in the set: "))
    s = []
    print("Enter the elements:")
    for _ in range(n):
        # Change int(input()) to input() if you want to work with strings.
        s.append(int(input()))
    
    print("\nAll subsets:")
    generate_subsets(s, 0, [])

File: Intermediate/30.Generate all subsets of a given set using recursion/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

// Recursive function to generate subsets
void generateSubsets(const vector<int>& set, int index, vector<int>& currentSubset) {
    // Base case: if we've processed all elements, print the current subset.
    if (index == set.size()) {
        cout << "{ ";
        for (int num : currentSubset)
            cout << num << " ";
        cout << "}" << endl;
        return;
    }

    // Recurse without including the current element.
    generateSubsets(set, index + 1, currentSubset);

    // Include the current element.
    currentSubset.push_back(set[index]);
    generateSubsets(set, index + 1, currentSubset);

    // Backtrack: remove the last element before returning.
    currentSubset.pop_back();
}

int main() {
    int n;
    cout << "Enter the number of elements in the set: ";
    cin >> n;

    vector<int> set(n);
    cout << "Enter the elements:" << endl;
    for (int i = 0; i < n; i++) {
        cin >> set[i];
    }

    vector<int> currentSubset;
    cout << "\nAll subsets:" << endl;
    generateSubsets(set, 0, currentSubset);

    return 0;
}
